docs/qos_enforcement.md

```diff
@@ -0,0 +1,254 @@
+# Magma QoS Enforcement
+
+*Status: In-Review*\
+*Author: @andreilee*\
+*Last Updated: 07/29*
+
+This document concerns:
+
+1. The addition of configurable QoS
+2. The addition of more comprehensive associations between subscribers, QoS, policies, and APN
+3. The enforcement of the above with an eventually consistent model
+
+
+## What We Want to Accomplish
+
+We would like to enable QoS enforcement, and move towards a model that more 
+closely enforces across UE level, APN level, and bearer level.
+
+**Private LTE**
+
+As an example use case for QoS enforcement, a service provider may wish to 
+segment users into tiers denoting their service quality, eg. bronze, silver, 
+and gold. A service provider should then be able configure a QoS profiles for 
+each of the three service levels. There should be the capability to define 
+which APNs support which QoS profiles, and which subscribers are allowed 
+which QoS levels, based on their subscriptions. Naturally, different policy 
+rules will also be tied to different QoS profiles, and together with APN 
+restrictions on supported QoS profiles, this places a restriction on which 
+policy rules may be installed for different users.
+
+Beyond segmenting subscribers into QoS levels, as our support for multi-APN
+matures, we wish to enable QoS enforcement for multiple sessions 
+of a single subscriber, one for each APN. Across these different sessions, each 
+APN will have a different QoS for its default bearer. To enable this behavior, 
+there should be the capability to assign a QoS profile to an APN for the 
+default bearer.
+
+It is also possible that a network operator would desire to have backhaul 
+specified differently for each APN even when served from the same gateway. 
+This requirement can arise naturally from prioritizing one service type, and 
+thus APN, over another. 
+We will define the `ApnServiceType` as a configuration specifying how backhaul 
+is set up for an APN in a gateway. 
+
+To enable the above, we have the following list of new API requirements for Magma:
+
+1. Configuration of QoS profiles
+2. Configuration of APN entities
+3. Configuration of ApnServiceType entities
+3. Assignment of QoS profiles to subscribers (to denote what QoS levels are allowed for the subscriber)
+4. Assignment of QoS profiles to APN entities (to denote what QoS levels are supported by the APN)
+5. Assignment of APN entities to subscribers (to denote which APNs a subscriber has access to)
+6. Assignment of ApnServiceType entities to gateways
+
+**Federated LTE**
+
+In federated use cases, we have slightly different requirements. APN entities 
+do not need to be configured as they are provided by the federation network.
+
+We also wish to support cases where an HSS is not provided, and our 
```

Comment:
For the case where our subscriberdb is used we will need to allow for APN entities to be specified by us. i.e. only when HSS is federated we disable our APN configuration.